src/cargo/core/compiler/compile_kind.rs

```diff
@@ -66,19 +67,34 @@ impl CompileKind {
                 .into_iter()
                 .collect());
         }
-        let kind = match &config.build_config()?.target {
-            Some(val) => {
-                let value = if val.raw_value().ends_with(".json") {
-                    let path = val.clone().resolve_path(config);
-                    path.to_str().expect("must be utf-8 in toml").to_string()
-                } else {
-                    val.raw_value().to_string()
-                };
-                CompileKind::Target(CompileTarget::new(&value)?)
+
+        let kinds = match &config.build_config()?.target {
+            None => vec![CompileKind::Host],
+            Some(build_target_config) => {
+                let values = build_target_config.values(config);
+                if values.len() > 1 && !config.cli_unstable().multitarget {
+                    bail!("specifying multiple `--target` flags requires `-Zmultitarget`")
+                }
```

Comment:
Can this be moved to `BuildTargetConfig.values`, and check if an array is used there? That way, this can guard against a single element array such as `build.target = ["foo"]`, which is not guarded against here.

Also, I think the text here could be adjusted to say it is for a config file value, not `--target`.